proto-public/pbmesh/v1alpha1/pbproxystate/traffic_permissions.proto

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+syntax = "proto3";
+
+package hashicorp.consul.mesh.v1alpha1.pbproxystate;
+
+message L7TrafficPermission {}
+
+enum TrafficPermissionAction {
+  INTENTION_ACTION_DENY = 0;
+  INTENTION_ACTION_ALLOW = 1;
+}
+
+message L4TrafficPermissions {
+  TrafficPermissionAction default_action = 1;
+  // permission is constructed by pre-flattening all relevant permissions.
+  repeated L4Permission permissions = 2;
+}
+
+message L4Permission {
+  repeated L4Principal allow_principals = 1;
+  repeated L4Principal deny_principals = 2;
+
+  // We don't need destination rules here because they either apply to L7 features or multi-ports.
+  // In the case of multiple ports, the xds controller is responsible for filtering per-port permissions.
+}
+
+// L4Principal maps into Source. We first convert this to Source before generating Envoy resources.
+message L4Principal {
+  string spiffe_regex = 1;
+  repeated string exclude_spiffe_regexes = 2;
+}
+
+message L7Principal {
```

Comment:
Since we're only doing L4 I wonder if we should omit it altogether. Otherwise, what do you think if we do the following:


```
message L7Principal {
  L7PrincipalID principal_id = 1;
  repeated L7PrincipalID exclude_principal_ids = 2;
}

message L7PrincipalID {
  oneof id {
    MTLSAuthPrincipal mtls = 1;
    XFCCPrincipal xfcc = 2;
    JWTPrincipal jwt = 3;
  }
}

message MTLSAuthPrincipal {
  string spiffe_regex = 1;
}

message XFCCPrincipal {
  string spiffe_regex = 1;
}

message JWTPrincipal {
  // todo:
}

message L7TrafficPermissions {
  repeated L7Policy allow_policies = 1;
  repeated L7Policy deny_policies = 2;
}

message L7Policy {
  repeated L7Principal principals = 1;
  repeated L7Rule rules = 2;
}

message L7Rule {
  L7PolicyRule rule = 1;
  repeated L7PolicyRule exclude_rules = 2;
}

message L7PolicyRule {
  string path_exact = 1;
  string path_prefix = 2;
  string path_regex = 3;

  // methods is the list of HTTP methods. If no methods are specified,
  // this rule will apply to all methods.
  repeated string methods = 4;

  RuleHeader header = 5;
}
```

The idea I have here with principals based on current usage is that it seems that today we use either `Authenticated`, or `Header` (only for xfcc) or `Segment` (only for JWT). Structuring it this way would allow us for more flexibility when we implement JWT without changing the API too much. 

The naming with L7 is hard so I ended up using the word `Policy` but instead of repeating `Permission` again.